File: WebSockets/MyRestaurant/app/src/main/java/com/emedinaa/myrestaurant/model/interactors/BaseRemoteInteractor.java
package com.emedinaa.myrestaurant.model.interactors;

import android.support.annotation.NonNull;

import com.emedinaa.myrestaurant.data.DataInjector;
import com.emedinaa.myrestaurant.data.callback.DataCallback;
import com.emedinaa.myrestaurant.data.remote.RestaurantEndPoint;

import retrofit2.Call;

/**
 * @author : Eduardo Medina
 * @see : https://futurestud.io/tutorials/retrofit-2-cancel-requests
 * @since : 8/10/18
 */
public class BaseRemoteInteractor<T> {

    protected final String DEFAULT_MSG_ERROR="Ocurrió un error";
    protected Call<T> currentCall=null;
    protected RestaurantEndPoint remote;

    public BaseRemoteInteractor() {
        remote= DataInjector.getInstance().remote().restaurantEndPoint();
    }

    protected void cancel(){
        if(currentCall!=null){
            currentCall.cancel();
        }
    }

    protected void processFailure(Call<T> call, Throwable t,
                                  @NonNull DataCallback dataCallback){
        if(call!=null){
            if(call.isCanceled()){ }else{
                dataCallback.onFailure(new Exception(t));
            }
        }
    }
}

File: WebSockets/ChefKotlinApp/app/src/main/java/com/kotlin/samples/kotlinapp/model/interactors/BaseRemoteInteractor.kt
package com.kotlin.samples.kotlinapp.model.interactors

import com.kotlin.samples.kotlinapp.data.DataInjector
import com.kotlin.samples.kotlinapp.data.callback.DataCallback
import com.kotlin.samples.kotlinapp.data.remote.KitchenEndPoint
import retrofit2.Call

open class BaseRemoteInteractor<T> {

    companion object {
        const val DEFAULT_MSG_ERROR:String="Ocurrió un error"
    }

    protected var currentCall:Call<T>?=null
    protected var remote: KitchenEndPoint = DataInjector.getInstance().remote().restaurantEndPoint()

    protected fun cancel() {
        currentCall?.let {
            it.cancel()
        }
    }

    protected fun processFailure(call: Call<T>?, t: Throwable,
                                 dataCallback: DataCallback) {
        if (call != null) {
            if (call.isCanceled) {
            } else {
                dataCallback.onFailure(Exception(t))
            }
        }
    }
}